src/components/tl/ucp/allreduce/allreduce_sliding_window.c

```diff
@@ -0,0 +1,368 @@
+    ucc_status_t       st        = UCC_OK;
+    ucc_status_t       global_st = ucc_ee_executor_finalize(
+                                    task->allreduce_sliding_window.executor);
+
+    st = ucc_tl_ucp_coll_finalize(&task->super);
+    if (ucc_unlikely(st != UCC_OK)) {
+        tl_error(UCC_TASK_LIB(task), "failed finalize collective");
+        global_st = st;
+    }
+
+    return global_st;
+}
+
+static inline void
+ucc_tl_ucp_allreduce_sliding_window_reduction(ucc_coll_task_t *coll_task,
+                                              ucc_tl_ucp_allreduce_sw_buf_t *accbuf,
+                                              ucc_tl_ucp_allreduce_sw_buf_t *getbuf)
+{
+    ucc_status_t       status = UCC_OK;
+    ucc_tl_ucp_task_t *task   = ucc_derived_of(coll_task, ucc_tl_ucp_task_t);
+    ucc_coll_args_t   *args   = &TASK_ARGS(task);
+    ucc_datatype_t     dt     = TASK_ARGS(task).dst.info.datatype;
+
+    status = ucc_dt_reduce(accbuf->buf,
+                           getbuf->buf,
+                           accbuf->buf,
+                           accbuf->count, dt,
+                           args, 0,
+                           0, task->allreduce_sliding_window.executor,
+                           &task->allreduce_sliding_window.etask);
+
+    if (ucc_unlikely(UCC_OK != status)) {
+        tl_error(UCC_TASK_LIB(task), "failed to perform dt reduction");
+        task->super.status = status;
+        return;
+    }
+
+    #define SAVE_STATE(_phase) do { } while(0) 
+    EXEC_TASK_TEST(1, "failed to perform dt reduction",
+                   task->allreduce_sliding_window.etask);
+}
+
+static inline ucs_status_t
+ucc_tl_ucp_allreduce_sliding_window_req_test(ucs_status_ptr_t request,
+                                             ucc_tl_ucp_task_t *task)
+{
+    if (request == NULL) {
+        return UCS_OK;
+    } else if (UCS_PTR_IS_ERR(request)) {
+        tl_error(UCC_TASK_LIB(task), "unable to complete UCX request=%p: %d\n",
+                 request,
+                 UCS_PTR_STATUS(request));
+        return UCS_PTR_STATUS(request);
+    } else {
+        return ucp_request_check_status(request);
+    }
+}
+
+static inline ucs_status_t
+ucc_tl_ucp_allreduce_sliding_window_req_wait(ucp_worker_h ucp_worker,
+                                             ucs_status_ptr_t request,
+                                             ucc_tl_ucp_task_t *task)
+{
+    ucs_status_t status;
+
+    if (request == NULL) {
+        return UCS_OK;
+    } else if (UCS_PTR_IS_ERR(request)) {
+        status = ucp_request_check_status(request);
+        tl_error(UCC_TASK_LIB(task), "unable to complete UCX request=%p: %d\n",
+                    request,
+                    UCS_PTR_STATUS(request));
+        return UCS_PTR_STATUS(request);
+    } else {
+        do {
+            ucp_worker_progress(ucp_worker);
+            status = ucp_request_check_status(request);
+        } while (UCS_INPROGRESS == status);
```

Comment:
can you make this nonblocking? this is against semantics

Reply:
I've removed it since it actually wasnt needed